In [1]:
import collections
import json
import re
import string

import nltk
import numpy as np
import pandas as pd
from bs4 import BeautifulSoup
from gensim.parsing.preprocessing import (preprocess_string, remove_stopwords,
                                          strip_punctuation, strip_tags)
from nltk.corpus import stopwords
from nltk.util import ngrams  # function for making ngrams
from numpy import asarray, save, savez_compressed
from sklearn.feature_extraction.text import CountVectorizer

In [2]:
import torch
torch.cuda.is_available()

False

In [3]:
mpstDF= pd.read_csv("mpst.csv")
mpstDF


imdb_id                                          title  \
0      tt0057603                        I tre volti della paura   
1      tt1733125  Dungeons &amp; Dragons: The Book of Vile Darkness   
2      tt0033045                     The Shop Around the Corner   
3      tt0113862                             Mr. Holland&#39;s Opus   
4      tt0086250                                       Scarface   
...          ...                                            ...   
14823  tt0219952                                  Lucky Numbers   
14824  tt1371159                                     Iron Man 2   
14825  tt0063443                                     Play Dirty   
14826  tt0039464                                      High Wall   
14827  tt0235166                               Against All Hope   

                                           plot_synopsis  \
0      Note: this synopsis is for the orginal Italian...   
1      Two thousand years ago, Nhagruul the Foul, a s...   
2      Matuschek&#39;s, a gift store in Budapest, is the ...   
3      Glenn Holland, not a morning person by anyone&#39;...   
4      In May 1980, a Cuban man named Tony Montana (A...   
...                                                  ...   
14823  In 1988 Russ Richards (John Travolta), the wea...   
14824  In Russia, the media covers Tony Stark&#39;s discl...   
14825  During the North African Campaign in World War...   
14826  Steven Kenet catches his unfaithful wife in th...   
14827  Sometime in the 1950s in Chicago a man, Cecil ...   

                                                    tags  split  \
0              cult, horror, gothic, murder, atmospheric  train   
1                                               violence  train   
2                                               romantic   test   
3                 inspiring, romantic, stupid, feel-good  train   
4      cruelty, murder, dramatic, cult, violence, atm...    val   
...                                                  ...    ...   
14823                                     comedy, murder   test   
14824                         good versus evil, violence  train   
14825                                           anti war  train   
14826                                             murder   test   
14827                                     christian film   test   

      synopsis_source  
0                imdb  
1                imdb  
2                imdb  
3                imdb  
4                imdb  
...               ...  
14823       wikipedia  
14824       wikipedia  
14825       wikipedia  
14826       wikipedia  
14827       wikipedia  

[14828 rows x 6 columns]

In [4]:
def decontracted(phrase):
    # specific
    phrase = re.sub(r"won't", "will not", phrase)
    phrase = re.sub(r"can't", "can not", phrase)
    phrase = re.sub(r"couldn't", "could not", phrase)
    phrase = re.sub(r"wouldn't", "would not", phrase)
    phrase = re.sub(r"shouldn't", "should not", phrase)
    phrase = re.sub(r"don't", "do not", phrase)
    phrase = re.sub(r"doesn't", "does not", phrase)
    phrase = re.sub(r"haven't", "have not", phrase)
    phrase = re.sub(r"hasn't", "has not", phrase)
    phrase = re.sub(r"ain't", "not", phrase)
    phrase = re.sub(r"hadn't", "had not", phrase)
    phrase = re.sub(r"didn't", "did not", phrase)
    phrase = re.sub(r"wasn't", "was not", phrase)
    phrase = re.sub(r"aren't", "are not", phrase)
    phrase = re.sub(r"isn't", "is not", phrase)

    # general
    phrase = re.sub(r"n\'t", " not", phrase)
    phrase = re.sub(r"\'re", " are", phrase)
    phrase = re.sub(r"\'s", " is", phrase)
    phrase = re.sub(r"\'d", " would", phrase)
    phrase = re.sub(r"\'ll", " will", phrase)
    phrase = re.sub(r"\'t", " not", phrase)
    phrase = re.sub(r"\'ve", " have", phrase)
    phrase = re.sub(r"\'m", " am", phrase)
    return phrase

# stop_words = stopwords.words('english')


In [5]:
mpstDF_processsed=mpstDF.copy()
# Type 1: Decontracted Text, The puncutation and stop words are still there
mpstDF_processsed["processed_synopsis_t1"]=mpstDF_processsed["plot_synopsis"].apply(lambda x: decontracted(" ".join(preprocess_string(x, [lambda x: x.lower(), strip_tags]))))
# Type 2 Decontracted Text Stop Words Removed
mpstDF_processsed["processed_synopsis_t2"]=mpstDF_processsed["plot_synopsis"].apply(lambda x: decontracted(" ".join(preprocess_string(x, [lambda x: x.lower(), strip_tags,remove_stopwords]))))

In [6]:
mpstDF_processsed

imdb_id                                          title  \
0      tt0057603                        I tre volti della paura   
1      tt1733125  Dungeons &amp; Dragons: The Book of Vile Darkness   
2      tt0033045                     The Shop Around the Corner   
3      tt0113862                             Mr. Holland&#39;s Opus   
4      tt0086250                                       Scarface   
...          ...                                            ...   
14823  tt0219952                                  Lucky Numbers   
14824  tt1371159                                     Iron Man 2   
14825  tt0063443                                     Play Dirty   
14826  tt0039464                                      High Wall   
14827  tt0235166                               Against All Hope   

                                           plot_synopsis  \
0      Note: this synopsis is for the orginal Italian...   
1      Two thousand years ago, Nhagruul the Foul, a s...   
2      Matuschek&#39;s, a gift store in Budapest, is the ...   
3      Glenn Holland, not a morning person by anyone&#39;...   
4      In May 1980, a Cuban man named Tony Montana (A...   
...                                                  ...   
14823  In 1988 Russ Richards (John Travolta), the wea...   
14824  In Russia, the media covers Tony Stark&#39;s discl...   
14825  During the North African Campaign in World War...   
14826  Steven Kenet catches his unfaithful wife in th...   
14827  Sometime in the 1950s in Chicago a man, Cecil ...   

                                                    tags  split  \
0              cult, horror, gothic, murder, atmospheric  train   
1                                               violence  train   
2                                               romantic   test   
3                 inspiring, romantic, stupid, feel-good  train   
4      cruelty, murder, dramatic, cult, violence, atm...    val   
...                                                  ...    ...   
14823                                     comedy, murder   test   
14824                         good versus evil, violence  train   
14825                                           anti war  train   
14826                                             murder   test   
14827                                     christian film   test   

      synopsis_source                              processed_synopsis_t1  \
0                imdb  note: this synopsis is for the orginal italian...   
1                imdb  two thousand years ago, nhagruul the foul, a s...   
2                imdb  matuschek is, a gift store in budapest, is the...   
3                imdb  glenn holland, not a morning person by anyone ...   
4                imdb  in may 1980, a cuban man named tony montana (a...   
...               ...                                                ...   
14823       wikipedia  in 1988 russ richards (john travolta), the wea...   
14824       wikipedia  in russia, the media covers tony stark is disc...   
14825       wikipedia  during the north african campaign in world war...   
14826       wikipedia  steven kenet catches his unfaithful wife in th...   
14827       wikipedia  sometime in the 1950s in chicago a man, cecil ...   

                                   processed_synopsis_t2  
0      note: synopsis orginal italian release segment...  
1      thousand years ago, nhagruul foul, sorcerer re...  
2      matuschek is, gift store budapest, workplace a...  
3      glenn holland, morning person anyone is standa...  
4      1980, cuban man named tony montana (al pacino)...  
...                                                  ...  
14823  1988 russ richards (john travolta), weatherman...  
14824  russia, media covers tony stark is disclosure ...  
14825  north african campaign world war ii, captain d...  
14826  steven kenet catches unfaithful wife apartment...  
14827  1950s chicago man, cecil moe (michael madsen) ...  

[14828 rows x 8 columns]

In [7]:
# from transformers import XLNetConfig, XLNetModel, XLNetTokenizer
from transformers import AutoModel,AutoConfig,AutoTokenizer


In [24]:
# xlnConfig= XLNetConfig()
# xlnModel = XLNetModel(xlnConfig)
from transformers import AutoModel,AutoConfig,AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained('xlnet-base-cased')
# model = AutoModel.from_config(config)
model = AutoModel.from_pretrained('xlnet-base-cased',mem_len=100)

In [9]:
def get_Encodings(text,tokenizer=tokenizer,model=model, verbose=False):
    if verbose:
        print("Text:")
        print(text[:20] + "....")
    # if the sequence lenght is too big we trim it to 250
    encoded = tokenizer.encode(text)
    
    if len(encoded)> 248:
        encoded = encoded[:248]+[4,3] # adding the encoding for special tokens by default
    if verbose:
        print("Encoded",encoded)
        print("length of Encoded", len(encoded))
    text_encoding_tensor=torch.tensor([encoded])
    if verbose:
        print("text_encoding_tensor:")
        print(text_encoding_tensor)
        print("shape:")
        print(text_encoding_tensor.shape)
    attention_mask_tensor= torch.tensor([[1]*text_encoding_tensor.shape[1]])
    if verbose:
        print("attention_mask_tensor:")
        print(attention_mask_tensor)
        print("shape:")
        print(attention_mask_tensor.shape)

    with torch.no_grad():
        outputs = model(text_encoding_tensor, attention_mask=attention_mask_tensor)
        if verbose:
            print("outputs:")
            print(outputs)
            print("Lenght of outputs",len(outputs))
            print("outputs[0]:")
            print(outputs[0])
            print("outputs[0].shape:")
            print(outputs[0].shape)
            print("outputs[1]:")
            print(outputs[1])
            print("Length Ooutputstput[1]:")
            print(len(outputs[1]))
            print("Sample from Output[1], first hidden layer:")
            print(outputs[1][0])
            print("Sample shape, first hidden layer")
            
    if verbose:
        print("getting the last tensor for XLNet")
        print(outputs[0].squeeze()[-1])
    return outputs[0].squeeze()[-1]

In [10]:
def getXLNetEmbeddings(df,column,verbose=False):
    embeddings = np.array(torch.tensor([np.array(get_Encodings(x,verbose=verbose)) for x in df[column]]))
    return embeddings

In [11]:
# THIS IS NORMAL TEST FUNCTION CREATED AND LATER MNERGED WITH THE getXLNetEmbeddings METHOD
# def getXLNetEmbeddings_testMode(df,column,verbose=False):
#     embeddings = np.array(torch.tensor([np.array(get_Encodings(x,verbose=verbose)) for x in df[column]]))
#     return embeddings

**Testing verbose mode for one Input**

In [65]:
sample_text="In May 1980, a Cuban man named Tony Montana (Al Pacino) claims asylum, in Florida, USA, and is in search of the \"American Dream\" after departing Cuba in the Mariel boatlift of 1980. When questioned by three tough-talking INS officials, they notice a tattoo on Tony's left arm of a black heart with a pitchfork through it, which identifies him as a hitman, and detain him in a camp called 'Freedomtown' with other Cubans, including Tony's best friend and former Cuban Army buddy Manolo \"Manny Ray\" Ribiera (Steven Bauer), under the local I-95 expressway while the government evaluates their visa petitions.After 30 days of governmental dithering and camp rumors, Manny receives an offer from the Cuban Mafia which he quickly relays to Tony. If they kill Emilio Rebenga (Roberto Contreras) a former aide to Fidel Castro who is now detained in Freedomtown, they will receive green cards. Tony agrees, and kills Rebenga during a riot at Freedomtown."
test_output=get_Encodings(sample_text)
print(test_output)
print(test_output.shape)


tensor([-2.1453e-01,  9.8638e-01, -1.3569e+00,  4.9744e-02, -2.0580e-01,
         2.0028e-01, -1.1084e+00, -2.4160e-01,  5.4918e-01,  6.1399e-01,
         2.2983e-01, -2.1084e+00,  4.5196e-01, -4.5227e-01,  1.9634e+00,
        -7.0501e-01,  7.6946e-01,  1.3087e+00,  4.6494e-01, -1.6070e+00,
         8.9788e-01, -1.1989e+00, -4.9769e-02, -7.3024e-01, -4.6916e-01,
         6.4692e-01, -1.2097e+00,  9.8949e-01, -1.8701e-01, -1.4202e+00,
         3.2042e-01, -2.4059e-02,  1.2414e+00,  7.3341e-02, -8.1281e-01,
        -9.1461e-01, -1.1381e+00,  3.1421e-01,  8.6016e-01,  1.2998e-02,
         5.2342e-01,  5.1991e-01, -1.0325e+00, -3.0562e-01, -2.7786e-01,
        -5.8735e-01,  1.1165e+00,  1.5297e+00, -6.8228e-01,  3.2196e-01,
        -3.0287e-01, -6.1153e-01, -1.8400e+00, -6.7133e-01,  6.4504e-01,
         7.3167e-01,  2.7735e-01, -1.7589e+00,  4.9479e-01,  1.1035e+00,
         2.3803e-03,  1.0375e+00, -2.1944e-01, -1.5443e+00,  4.3127e-01,
         1.3444e+00, -1.4885e+00,  4.8591e-01,  6.0

In [72]:

def truncateText(text,size):
    return (" ".join(sample_text.split(" ")[:size]))



[&#39;In&#39;, &#39;May&#39;, &#39;1980,&#39;, &#39;a&#39;, &#39;Cuban&#39;, &#39;man&#39;, &#39;named&#39;, &#39;Tony&#39;, &#39;Montana&#39;, &#39;(Al&#39;, &#39;Pacino)&#39;, &#39;claims&#39;, &#39;asylum,&#39;, &#39;in&#39;, &#39;Florida,&#39;, &#39;USA,&#39;, &#39;and&#39;, &#39;is&#39;, &#39;in&#39;, &#39;search&#39;, &#39;of&#39;, &#39;the&#39;, &#39;&quot;American&#39;, &#39;Dream&quot;&#39;, &#39;after&#39;, &#39;departing&#39;, &#39;Cuba&#39;, &#39;in&#39;, &#39;the&#39;, &#39;Mariel&#39;, &#39;boatlift&#39;, &#39;of&#39;, &#39;1980.&#39;, &#39;When&#39;, &#39;questioned&#39;, &#39;by&#39;, &#39;three&#39;, &#39;tough-talking&#39;, &#39;INS&#39;, &#39;officials,&#39;, &#39;they&#39;, &#39;notice&#39;, &#39;a&#39;, &#39;tattoo&#39;, &#39;on&#39;, &quot;Tony&#39;s&quot;, &#39;left&#39;, &#39;arm&#39;, &#39;of&#39;, &#39;a&#39;, &#39;black&#39;, &#39;heart&#39;, &#39;with&#39;, &#39;a&#39;, &#39;pitchfork&#39;, &#39;through&#39;, &#39;it,&#39;, &#39;which&#39;, &#39;identifies&#

Now testing the helper functions for input from a dataframe to see if we get the correct desired output for our work

In [13]:
sampleDF=mpstDF_processsed[["processed_synopsis_t1","processed_synopsis_t2"]].head(3)
sampleDF["processed_synopsis_t1"]=sampleDF["processed_synopsis_t1"].apply(lambda x : x[:1000])
sampleDF["processed_synopsis_t2"]=sampleDF["processed_synopsis_t2"].apply(lambda x : x[:1000])
display(sampleDF)

processed_synopsis_t1  \
0  note: this synopsis is for the orginal italian...   
1  two thousand years ago, nhagruul the foul, a s...   
2  matuschek is, a gift store in budapest, is the...   

                               processed_synopsis_t2  
0  note: synopsis orginal italian release segment...  
1  thousand years ago, nhagruul foul, sorcerer re...  
2  matuschek is, gift store budapest, workplace a...

In [14]:
# sample_embeddings=getXLNetEmbeddings_testMode(sampleDF,"processed_synopsis_t1",verbose=True)
sample_embeddings=getXLNetEmbeddings(sampleDF,"processed_synopsis_t1")
print("sample_embeddings.shape",sample_embeddings.shape)
print(sample_embeddings)


sample_embeddings.shape (3, 768)
[[-1.5467558   0.41237524 -1.3485138  ... -1.2315857   0.12427384
  -0.17078479]
 [-0.7172874   0.66564727  0.51811284 ... -0.56954485  2.5583036
  -1.0299408 ]
 [-1.7686011  -0.73729324 -1.3424073  ... -0.13613753 -0.6869708
   0.5819778 ]]


HERE: make sure that the output is (3, 768). Where 3 is the number of texts we sent to the model and 768 is the output of embedding length for the model chosen.

Now We run the model for each text we have in our Dataset 

We send the data in batches

In [32]:
# for t1 first:
def get_data_in_batch(df, col_name, total_length, batch_size):
    for i in range(0,total_length,batch_size):
        start=i
        end= min(i+batch_size,total_length)
        print("{} to {}".format(start,end))
        batch= df[col_name][start:end].to_list()
        print(batch)
        Encodings=tokenizer.encode([batch])
        print(Encodings)
        break

In [ ]:
def getXLNetEmbeddingsForBatch(batch,verbose=False):
    
    embeddings = np.array(torch.tensor([np.array(get_Encodings(x,verbose=verbose)) for x in df[column]]))
    return embeddings

In [33]:
get_data_in_batch(mpstDF_processsed,"processed_synopsis_t1",14828,10)

0 to 10
[&#39;note: this synopsis is for the orginal italian release with the segments in this certain order.boris karloff introduces three horror tales of the macabre and the supernatural known as the  nothree faces of fear\&#39;.the telephonerosy (michele mercier) is an attractive, high-priced parisian call-girl who returns to her spacious, basement apartment after an evening out when she immediately gets beset by a series of strange phone calls. the caller soon identified himself as frank, her ex-pimp who has recently escaped from prison. rosy is terrified for it was her testimony that landed the man in jail. looking for solace, rosy phones her lesbian lover mary (lynda alfonsi). the two women have been estranged for some time, but rosy is certain that she is the only one who can help her. mary agrees to come over that night. seconds later, frank calls again, promising that no matter who she calls for protection, he will have his revenge. unknown to rosy, mary is the caller imperson

ValueError: Input [[&#39;note: this synopsis is for the orginal italian release with the segments in this certain order.boris karloff introduces three horror tales of the macabre and the supernatural known as the  nothree faces of fear\&#39;.the telephonerosy (michele mercier) is an attractive, high-priced parisian call-girl who returns to her spacious, basement apartment after an evening out when she immediately gets beset by a series of strange phone calls. the caller soon identified himself as frank, her ex-pimp who has recently escaped from prison. rosy is terrified for it was her testimony that landed the man in jail. looking for solace, rosy phones her lesbian lover mary (lynda alfonsi). the two women have been estranged for some time, but rosy is certain that she is the only one who can help her. mary agrees to come over that night. seconds later, frank calls again, promising that no matter who she calls for protection, he will have his revenge. unknown to rosy, mary is the caller impersonating frank. marry arrives at rosy is apartment soon after, and does her best to calm rosy is nerves. she gives the panic-struck woman a tranquillizer and puts her to bed.later that night as rosy sleeps, mary gets up out of bed, and pens a note of confession: she was the one making the strange phone calls when she learned of franks escape from prison. knowing that rosy would call on her for help, she explains that she felt it was her way of coming back into her life after their breakup. while she is busy writing, she fails to notice an intruder in the apartment. this time it is frank, for real. he creeps up behind mary and strangles her to death with one of rosys nylon stockings. the sound of the struggle awaken rosy and she gasps in fright. the murderous pimp realizes that he just killed the wrong woman, and slowly makes his way to rosy is bed. however, earlier that night, rosy had placed a butcher knife under her pillow at mary is suggestion. rosy seizes the knife and stabs frank with it as he is beginning to strangle her. rosy drops the knife and breaks down in hysteria, surrounded by the two corpses of her former lovers.the wurdalakin 19th century russia, vladimir d\&#39;urfe is a young nobleman on a long trip. during the course of his journey, he finds a beheaded corpse with a knife plunged into its heart. he withdraws the blade and takes it as a souvenir.later that night, vladimir stops at a small rural cottage to ask for shelter. he notices several daggers hanging up on one of the walls, and a vacant space that happens to fit the one he has discovered. vladimir is surprised by the entrance of giorgio (glauco onorato), who explains that the knife belongs to his father, who has not been seen for five days. giorgio offers a room to the young count, and subsequently introduces him to the rest of the family: his wife (rika dialina), their young son ivan, giorgio is younger brother pietro (massimo righi), and sister sdenka (susy anderson). it subsequently transpires that they are eagerly anticipating the arrival of their father, gorcha, as well as the reason for his absence: he has gone to do battle with the outlaw and dreaded wurdalak ali beg. vladimir is confused by the term, and sdenka explains that a wurdalak is a walking cadaver who feeds on the blood of the living, preferably close friends and family members. giorgio and pietro are certain that the corpse vladimir had discovered is that of ali beg, but also realize that there is a strong possibility that their father has been infected by the blood curse too. they warn the count to leave, but he decides to stay and await the old mans return.at the stroke of midnight, gorcha (boris karloff) returns to the cottage. his sour demeanor and unkempt appearance bode the worse, and the two brothers are torn: they realize that it is their duty to kill gorcha before he feeds on the family, but their love for him makes it difficult to reach a decision. later that night, both ivan and pietro are attacked by gorcha who drains them of blood, and then flees the cottage. giorgio stakes and beheads pietro to prevent him from reviving as a wurdalak. but he is prevented from doing so to ivan when his wife threatens to commit suicide. reluntantly, he agrees to bury the child without taking the necessary precautions.that same night, the child rises from his grave and begs to be invited into the cottage. the mother runs to her son is aid, stabbing giorgio when he attempts to stop her, only to be greeted at the front door by gorcha. the old man bits and infects his daughter-in-law, who then does the same for her husband. vladimir and sdenka flee from the cottage and go on the run and hide out in the ruins of an abandoned cathedral as dawn breaks. vladimir is optimistic that a long and happy life lies with them. but sdenka is reluctant to relinquish her family ties. she believes that she is meant to stay with the family.sdenka is fears about her family are confirmed when that evening, gorcha and her siblings show up at the abandoned abby. as vladimir sleeps, sdenka is lured into their loving arms where they bite to death. awakened by her screams, vladimir rushes to her aid, but the family has already taken her home, forcing the lover to follow suite. the young nobleman finds her, lying motionless on her bed. sdenka awakens, and a distinct change is visible on her face. no longer caring, vladimir embraces her, and she bites and infects him too.the drop of waterin victorian london, england, nurse helen chester (jacqueline pierreux) is called to a large house to prepare the corpse of an elderly medium for her burial. as she dressed the body, she notices an elaborate diamond ring on its finger. tempted by greed, nurse chester steals it. as she does, a glass tips over, and drops of water begin to splash on the floor. she is also assailed by a fly, no doubt attracted by the odor of the body. unsettled but pleased by her acquisition, she finishes the job and returns home to her small east end flat.after returning home, nurse chester is assailed by strange events. the buzzing fly returns and continues to pester her. then the lights in her apartment go out, and the sounds of the dripping water continues with maddening regularity. she sees the old womans corpse lying on her bed, and coming towards her. the terrified woman begs for forgiveness, but she ultimately strangles herself, imaging that the medium is hands are gripping her throat.the next morning, the concierge (harriet white medin) discovers nurse chester is body and calls the police. the investigator on the scene (gustavo de nardo) quickly concludes that its a simple case and that nurse chester &quot;died of fright&quot;. the pathologist arrives on the scene to examine the body before it is taken away and he notes that the only sign of violence is a small bruise on her left finger, mostly likely caused when someone pried a ring from her finger. as the doctor makes this observation, the concierge appears distressed, as she has apparently took the ring from the dead nurse chester, and is further distracted by the sound of a fly swooping about in the air....boris karloff makes a final appearance as gorcha riding on his horse as he concludes the three tales of fear and tells the viewers to be careful while walking home at night for ghosts and vampires have no fear. the image pulls back to actually reveal him sitting on a prop fake horse with a camera crew and various crewmen moving branches around to simulate the scene of riding through the forest from the wurdalak segment.&#39;, &#39;two thousand years ago, nhagruul the foul, a sorcerer who reveled in corrupting the innocent and the spread of despair, neared the end of his mortal days and was dismayed. consumed by hatred for the living, nhagruul sold his soul to the demon lords of the abyss so that his malign spirit would survive. in an excruciating ritual, nhagrulls skin was flayed into pages, his bones hammered into a cover, and his diseased blood became the ink to pen a book most vile. creatures vile and depraved rose from every pit and unclean barrow to partake in the fever of destruction. the kingdoms of karkoth were consumed by this plague of evil until an order of holy warriors arose from the ashes. the knights of the new sun swore an oath to resurrect hope in the land. the purity of their hearts was so great that pelor, the god of light, gave the knights powerful amulets with which to channel his power. transcendent with divine might, the knights of the new sun pierced the shadow that had darkened the land for twelve hundred years and cast it asunder. but not all were awed by their glory. the disciples of nhagruul disassembled the book and bribed three greedy souls to hide the pieces until they could be retrieved. the ink was discovered and destroyed but, despite years of searching, the cover and pages were never found. peace ruled the land for centuries and the knights got lost in the light of their own glory. as memory of the awful events faded so did the power of servants of pelor. they unwittingly abandoned themselves in the incorrect belief that the book of vile darkness could never again be made whole.now, the remaining pieces have been discovered, and an ancient evil is attempting to bring them together and restore the relic and the evil it brought. but at the same time a potential new paladin has been named to the knights of the new sun to attempt to renew their power to fight this evil. but, to do so, he may need to go against all that he has held dear, risking more that just his own soul in his quest to destroy the evil that surrounds him at every turn.&#39;, &#39;matuschek is, a gift store in budapest, is the workplace of alfred kralik (james stewart) and the newly hi ed klara novak (margaret sullavan). at work they constantly irritate each other, but this daily aggravation is tempered by the fact that each has a secret pen pal with which they trade long soul-searching letters. romantic correspondence is sent back and forth, and while alfred and klara trade barbs at work, they dream of someday meeting their sensitive, caring and unknown pen pal.christmas is fast approaching, and the store is busy. alfred had been with the store for some time, and has always been treated well by mr. matuschek (frank morgan), but lately his attitude has changed. alfred is at a loss, and matuschek avoids any explanation, finally telling alfred that it would be best if he left. stunned, alfred accepts his last paycheck and says goodbye to everyone, including klara. for once they are civil to each other.a long awaited meeting of the secret pen pals was planned for that night, and alfred having just lost his job has no desire to go. finding he can not fight his curiosity, he wanders to the restaurant where they would agreed to meet and peeks in the window with his fellow employee. of course, klara is there waiting for him, with the chosen book and wearing a red carnation they would agreed to use as a signal. realizing that he would been wrong about her all along, and that his irritation with her was actually masking his attraction, he finally enters and goes over to her table, but does not reveal his true reason for being there although he is aware she will be hurt that her pen pal does not show up. alfred, hurt by her rudeness, finally leaves, knowing that she will wait all night for someone who is no longer coming.meanwhile, back at the store, mr. matuschek has a late-night meeting with a private detective. he knows that his wife has been having an affair with one of his employees, and was convinced it was his trusted friend, alfred. the detective however tells matuschek it is in fact another employee and, heart-broken over his wife is infidelity he retires to his office. the delivery boy, returning late, enters and prevents matuschek shooting himself with a pistol. collapsing in grief and shame, matuschekis rushed to the hospital.the next day alfred visits mr. matuschek in his sick bed, where he asks for alfred is forgiveness and puts him back to work, now as manager of the store. the delivery boy is rewarded with a raise to a store clerk. klara arrives at work late, obviously heartbroken after the failure of her correspondent to materialize last night. when she finds alfred in the manager is office she does not believe him and when she discovers it is true she faints in the middle of his office. later, as she is resting at home, alfred pays her a visit, and while he is there her aunt brings her another letter from her secret pen pal that explains his not being at the meeting because he saw her there with alfred. relieved about the misunderstanding she swears to alfred she will be back at work in the morning. alfred is obviously working on a plan to reveal himself to klara.christmas eve is here, and everyone works through the day. mr. matuschek has nearly recovered from his sickness and stops by to see how things are going, and when the final tally is made, the store has had its best sales day since 1928. delighted, he hands out bonuses to all, and takes the new stock boy out for christmas dinner. alfred and klara are getting ready to leave, and she has another date with her mystery pen pal, but alfred delays her with a few questions. she is never yet seen him and does not even know his name but is convince end she will be engaged when she comes back to work. he tells her that the mysterious pen pal stopped by to see him earlier, and he is in fact fat, bald, older and unemployed and quite willing to live off klara is income. alfred reveals himself when he puts a red carnation in his lapel and suddenly eveything becomes clear to her.&#39;, &#39;glenn holland, not a morning person by anyone is standards, is woken up by his wife iris early one bright september morning in 1964. glenn has taken a job as a music teacher at the newly renamed john f. kennedy high school. he intends his job to be a sabbatical from being a touring musician, during which he hopes to have more &quot;free time&quot; to compose. however, he soon finds that his job as a teacher is more time-consuming than he first thought.as he arrives at the school for the first time, he meets vice principal wolters, who comments on his corvair, the model of car the ralph nader wrote a book about. inside the building, he meets principal helen jacobs. having got off to an awkward start with both of them, he goes to the music room and meets his students for the first time. the students are dull, apathetic, and mostly terrible musicians. at lunchtime, he meets the football coach, bill meister, and strikes up a friendship with him. at the end of his stressful first day, glenn and iris talk about their future. if everything goes according to plan, between his paychecks and what she made with her photography, he should be able to quit in four years and go back to his music, including composing.glenn notices one dedicated but inept clarinet player, gertrude lang, and starts working with her individually. he continues attempting to teach the class about music and continues working on his music at home as time passes. grading papers gradually replaces working on his own music during his home time, much to his chagrin. after several months, glenn grows exasperated when it seems that none of his students have learned anything from his classes. gertrude, despite diligent practice, does not improve her clarinet-playing. glenn is exasperation is further compounded when the principal jacobs chastises him for not focusing properly on his students. she has noticed that he is even happier to leave at the end of each day than most of the students. later, glenn expounds his frustration to iris, who then informs him that she is pregnant. glenn is dumbstruck, and his muteness upsets iris. to comfort her, he tells her a story about how he discovered john coltrane (his favorite musician) records as a teenager, the point being he could get used to this turn of affairs.after some soul-searching, glenn decides to try some unconventional methods of teaching music appreciation, including the use of \&#39;rock and roll\&#39; to interest students, demonstrating to them the similarities between bach is &quot;minuet in g&quot; and rock-and-roll in the form of the toys\&#39; &quot;lovers concerto&quot;. for the first time, the students are interested in the class, and glenn appears much happier as he relates this to iris as they assemble a crib. their apartment is getting more and more crowded, and glenn suggests that they get a house. iris is overjoyed, even though it means using their savings and glenn sacrificing his summer vacation, which he intended to use to work on his composing, in order to make extra money teaching driver is ed. glenn does right by his family but he knows he can forget about getting out of the teaching gig for the foreseeable future.continuing his new, unorthodox teaching methods, he finally gets gertrude, who was on the verge of giving up, to have a breakthrough and become a more skilled clarinet player. she rediscovers her joy of playing, and the now-competent band go on to play at the 1965 graduation. summer vacation begins, and glenn follows through on his plan to teach driver is ed, having a series of near-death experiences at the hands of new drivers. glenn and iris move into their new house. soon, we see the driver is ed car once again, except this time it is glenn himself driving like a maniac, breaking every traffic law - so that he could get to the hospital to see his newborn son, coltrane (&quot;cole&quot;) holland.glenn is unorthodox teaching methods do not go unnoticed by principal jacobs, or vice-principal wolters. they, along with the conservative school board and the parents of the community, are hostile to rock-and-roll. glenn is able to convince the principal that he believes strongly that teaching the students about all music, including rock-and-roll, will help them appreciate it all the more. the principal and vice principal also hand him a new assignment, to get a marching band together for the football team. glenn is at a loss with this concept, until his bill meister agrees to help, in exchange for glenn putting one of his football players, louis russ, in the band to allow him to earn an extra curricular activity credit, which he needs in order to stay eligible for the sports teams. louis knows absolutely nothing about music, but takes up drums. he has trouble keeping time and always finds himself out of place with the rest of the band.later, glenn and bill are chatting while playing chess. bill, a bachelor, wants to know about glenn is stories of debauchery as a traveling musician, but glenn does not want to talk about the past, as he is a different person in a different time. glenn instead tells bill he is pessimistic about louis russ. bill encourages him to keep trying. much as he worked with gertrude earlier, glenn starts working one-on-one with louis, helping to get a feel for the tempo of music. after some hard work, louis gets it, and later, he marches with the band in the local parade, much to the delight of his family.immediately behind the kennedy band in the parade is a fire engine, and its deafeningly loud horn catches everyone by surprise. iris looks into cole is stroller to check on him, but the noise has not awakened cole - the boy is deaf. the revelation drives a wedge between glenn and his son, as it seems that his son cannot understand what he does. a more somber glenn teaches his students about beethoven, the deaf composer.time passes, and we see a montage of events from the late \&#39;60s, as glenn picks away as his composition a little at a time and watches iris work with cole. we stop again in the early \&#39;70s, with glenn still directing his high school band. cole is old enough to enter school. because of her mounting frustration with her inability to communicate with cole, she insists on sending him to a special school for the deaf, whatever the cost. the three of them visit the school. glenn winces at the cost, but they enroll cole and set about to learn sign language themselves, though iris puts more effort into it than glenn.apathetic students still go through glenn is classes, and one of them, named stadler, is stoned. glenn is chewing him out when glenn receives bad news. he tells stadler to meet him on saturday. on that day, they appear at a funeral. louis russ has come home from vietnam after being killed in action. coach meister is there, and he and glenn mourn. at the end of that academic year, bill reveals that he finally has a steady girl-friend, and principal jacobs retires from the high school, praising glenn for what he has done.we see another montage of events, this time in the 1970s. glenn continues teaching driver is ed in the summer. we see the class of 1980 being welcomed back, suggesting that it is now september 1979. glenn and bill meister team up to help the drama department, when it is rumored that funding may be pulled. glenn and bill tell wolters, now the principal, have an idea to be certain the school will make money rather than lose it; it will be a musical revue of gershwin classics. during auditions for the musical revue, glenn becomes entranced and interested in a talented young singer named rowena morgan. at home, the teenage cole comes home and tells glenn about the science fair, which glenn missed. iris is fluent at sign language now, but glenn is still only fair. iris reproaches him for spending so much time with the school projects and the students while neglecting his own son. glenn is frustrated, realizing that his own musical composing has been on the back burner for 15 years now.rowena visits glenn at a diner, where he has gotten into the habit of going to get out of the house and have someplace quiet to work. unknown to iris, glenn writes a small piece that he titles &quot;rowena is theme,&quot; and takes an interest when rowena states that she wants to leave town and go to new york to sing professionally. glenn is life at home is still strained. iris agrees to come to the school play on saturday, because she had a meeting with cole is teachers on opening night (friday).the school revue arrives at last, and is a big hit, playing to a packed house. in the audience we see coach meister wearing a ring (he married the woman we saw earlier), and sarah, the drama teacher, shows principal wolters something on a new invention, a handheld calculator (presumably, showing him how much gate money made it into the school coffers, as wolters looks impressed). after the revue, rowena comes to see glenn in the auditorium, and she tells him she intends to pursue her dream of singing by going to new york the very next night, after the second and last performance of the revue. glenn is taken aback. rowena hints that she would like glenn to come with her. glenn goes home and looks at his photo album, looking at pictures of his family and pictures of his old life as a traveling musician, now half a lifetime in the past. he is tempted to leave everything behind and go with rowena to restore his old life as a musician. however, he realizes he is no longer the same person as he was then. he visits rowena at the bus stop and sees her off, giving her the names of someone in new york who will help her find lodging. glenn watches her depart, and goes home, content in his love for iris.the timeline then shifts to late 1980, when john lennon is killed. glenn goes home and finds cole working on glenn is old corvair. when cole asks what is wrong, glenn tries to explain, but then gives up, feeling that his son would not understand john lennon or his music. this infuriates cole who (through iris), explains that he does care about glenn and knows about john lennon, but that glenn does not seem to be at all interested in communicating with him. cole berates his father for putting so much effort in to teaching his students and very little towards him, calling him an asshole in sign language as he stalks off. glenn then makes an effort, and even provides a concert at the high school, which also features lights and other items to enhance the show for deaf members of the school where cole attends. glenn, having become somewhat more proficient in sign language, even does an interpretation of lennon is song \&#39;beautiful boy,\&#39; dedicated to cole. later, glenn discovers cole listening to records by sitting on the speakers and feeling the vibrations through his body, and they can start healing the rift between them, even as glenn is composition continues to gather dust.time passes. it is now 1995. glenn goes to see principal wolters, who announces that art, music and drama have been cut from the school curriculum, and glenn would be out of a job shortly. glenn, who has become a cynical old man, tells wolters that to cut the fine arts would lead to a generation of students who would be proficient at reading and writing and math (maybe) but would have nothing to read or write about. wolters offers to write glenn a reference, but glenn, who is now 60 years old, fully recognizes the futility of the gesture. his working days are over and he knows it. then he looks up at the picture on the wall of the long-departed former principal jacobs. he says jacobs would have fought the budget cuts, and he will too. glenn pleads to the school board to reconsider, but they refuse.at home, iris reads a letter from the now-grown cole. he has become a teacher himself, and was considering an offer from a university for the deaf in washington, d.c. he also has taken glenn is old car, the corvair that we saw at the beginning and that cole was working on in his teens, and jokingly writes that he will never give it back. despondent, glenn walks through the school on his last day, and he talks to coach bill, whose job as football coach is safe, though he can not be far from retirement himself. glenn figures that he will bring in some money teaching piano lessons on the side, but he is unprepared to be forced into early retirement.on glenn is final day at the school, cole shows up driving the corvair. school is out for him, too. glenn is surprised when iris and cole lead him to the school auditorium, where they have organized a surprise going-away celebration for him. he sees many of his former students in the audience, including stadler, the pothead from years before. arriving next is gertrude lang, the clarinetist who glenn helped in the 60s, who has since become the state is governor. gertrude thanks glenn for his dedication, and glenn is very moved. he is moved to tears when she gives him a baton and asks him to conduct his own composition, which she had got hold of. the curtains open and a band, filled with more of glenn is former students, is assembled and ready to play. governor lang picks up her clarinet and takes her place among them, and they play, for the first time, the musical opus that glenn had been picking away at for three decades.&#39;, &#39;in may 1980, a cuban man named tony montana (al pacino) claims asylum, in florida, usa, and is in search of the &quot;american dream&quot; after departing cuba in the mariel boatlift of 1980. when questioned by three tough-talking ins officials, they notice a tattoo on tony is left arm of a black heart with a pitchfork through it, which identifies him as a hitman, and detain him in a camp called \&#39;freedomtown\&#39; with other cubans, including tony is best friend and former cuban army buddy manolo &quot;manny ray&quot; ribiera (steven bauer), under the local i-95 expressway while the government evaluates their visa petitions.after 30 days of governmental dithering and camp rumors, manny receives an offer from the cuban mafia which he quickly relays to tony. if they kill emilio rebenga (roberto contreras) a former aide to fidel castro who is now detained in freedomtown, they will receive green cards. tony agrees, and kills rebenga during a riot at freedomtown. the murder of rebenga was requested by frank lópez, a wealthy, politically astute man who deals cars and trades in cocaine, as rebenga had tortured lópez is brother to death while still in cuba many years earlier.after getting their green cards, tony montana and manny ray find work as dishwashers in a corner sandwich/taco shop. some weeks later, a lópez henchman and underboss, omar suárez (f. murray abraham), the man who contacted manny for the rebenga hit job, offers tony and manny a low-risk job of unloading marijuana from a boat from mexico to arrive in miami the following night for $500 each. tony insults suárez by turning down the job over the little money they will receive, and demands at least $1,000 for the work. after an altercation, suárez sets tony up on another job to purchase two kilograms of cocaine worth around $25,000 a piece from a colombian dealer, named hector the toad, a medium to high-risk job for which tony and manny will receive $5,000 for their work.that weekend, tony, manny, and two other marielitos in his crew whom they met in freedomtown, angel fernández (pepe serna), and chi chi (ángel salazar) then set out to meet &quot;hector the toad&quot; (al israel) at a seedy motel on the boulevard in miami beach. while manny and chi chi wait in the car on the street, tony and angel go up to the hotel room to meet with hector. the meeting does not go smoothly, as tony grows irritated with hector, who is slow to give him the cocaine in exchange for money. suddenly, tony and angel are double-crossed by the colombian. it becomes apparent that hector does not intend to sell tony the cocaine he has; he only wants to steal the money tony has been given to purchase the product. to convince tony to give over the cash, hector dismembers angel in a shower stall with a chainsaw. after angel is dead, tony, about to suffer the same fate, is saved by chi chi and manny who arrive in the nick of time to gun down hector is henchmen. manny receives a minor bullet wound in his shoulder when his uzi sub-machine gun jams. hector escapes but tony vengefully confronts him in the street and shoots him dead in the middle of the crowded ocean drive (the now famous miami south beach boulevard). tony and his crew then get away with both the cocaine and the money before the police arrive.the following night, tony and manny meet lópez (robert loggia) at his house for the first time where tony impresses lopez with not only the return of his cash but with a gift of the cocaine, a prize from the botched rip off. frank immediately hires tony and his crew into his criminal hierarchy, a representative of a cuban mafia. but during this initial get together tony also meets lopez is lady, the blond and beautiful elvira hancock (michelle pfeiffer), who will eventually become the source of tension between the two men. taking tony and manny out to a local nightclub, called the babylon club where frank frequently attends, tony and manny see first-hand the high standard of living they have come to acquire. though frank actually warns against these excesses, tony is seduced by them regardless. thus, tony montana begins his rise through the ranks of the miami cocaine underworld.three months later. tony montana attempts to make amends by meeting with his estranged family. it is implied that tony is father, a former u.s. navy sailor, abandoned the family when tony was little. since then, his mother (miriam colon) and younger 19-year-old sister gina (mary elizabeth mastrantonio) have been living in miami. tony shows up at his mother is and gina is house one evening, fashionably dressed, and offers them $1,000 in cash for financial support. gina is overjoyed to see her older brother whom they have not seen for five years. but tony is mother has only scorn for him since he turned his back on them many years ago for the quick and easy life of crime back in cuba, and wants nothing to do with tony, and she is too full of pride to accept his money despite being financially stricken. but gina, who idolizes her brother, follows him outside where he slips her the money secretly. gina tells tony that she wants in on the flashy life that he has going for him. tony is love for gina is clearly genuine for she is the only person that he trusts, and is also very protective of her. afterwords, manny makes a comment to tony about how attractive gina is, but tony angrily warns him to avoid courting her.several months later, tony is sent to bolivia to help omar set up a new distribution deal with bolivian kingpin alejandro sosa (paul shenar), since frank is having legal troubles that preclude him from leaving the country. though tony was supposed to let omar do all the talking, omar proves to be a poor negotiator, prompting tony to step in and save the deal. they seem to negotiate a deal that, on the surface seems favorable to both sides, but omar insists that frank would not approve. sosa then sides with omar on this and suggests that omar use his phone to call frank. a few minutes later, sosa hands tony binoculars, and he sees two menacing assassins, alberto the shadow (mark margolis) and the skull (geno silva), execute omar by hanging him by the neck from an airborne helicopter. sosa reveals that alberto recognized omar as once being an informant for the police several years ago, and he has a zero tolerance for disloyalty. tony insists that he never goes back on his word, and that he never trusted omar. sosa agrees to bring tony on board with him as his north american distributor of cocaine and other drugs. but upon their agreement, sosa sternly warns tony never to betray or double-cross him in any way.upon his return to florida, tony is chewed out by frank for overstepping his authority as well as hearing about omar is death. tony explains to frank that for a price of $18 million to pay sosa for the manufacturing and transportation costs, they will receive 2,000 kilograms of cocaine from bolivia for nationwide sale and distribution which will earn them $75 million over a period of one year. frank is worried because he does not have the many millions to pay sosa for the cocaine, but tony says that he is in tight with sosa and he has established a &quot;credit line&quot; with him as well as work out a payment plan where they will pay sosa $5 million up front and the rest in monthly installments. plus, in case frank comes up short a few million, tony will earn the money needed through his own street contacts. frank angrily tells tony that he did not negotiate a good deal, for sosa merely tricked him into thinking he did. tony replies that it is time for them to &quot;think big,&quot; and to expand the cartel for nationwide distribution of cocaine. with them as the main north american distributors and wholesalers of the sosa cartel, they will make millions and become the biggest cartel in the continent. frank warns tony that sosa cannot be trusted and that he will sooner or later turn against them for any slight deviation or compromise of his business. frank orders tony to stall his deal with sosa for the time being. frank then promptly tells tony that ambitious drug dealers, such at himself, who want too much and crave money, power and attention, do not last long in the business. tony leaves shrugging with indifference and strikes out on his own.(note: frank lópez is two warnings about tony is greed and sosa is violence will be later proven true.)after this incident, tony then seeks out elvira to whom he makes an unexpected marriage proposal. she is shaken by this, but agrees to think about it. frank lópez is none too happy when he hears about this and decides to take out tony.at the babylon nightclub that evening, tony is approached and shaken down by a miami police detective, named mel bernstein (harris yulin). he proposes to &quot;tax&quot; tony on his transactions in return for police protection and information. tony is distracted by the sight of gina dancing with a local drug dealer. he follows the two to a restroom stall where he berates gina for her promiscuous conduct. he asks manny to take her home. on the way gina admits she is attracted to manny. manny wards her off, mindful of tony is extreme protectiveness.back at the nightclub, tony is attacked by two gunmen but manages to escape, killing them both despite being wounded by a gunshot to his left shoulder. suspecting frank sent bernstein and the hitmen, tony asks one of his bodyguards, nick the pig, to call frank after tony arrives at frank is office at 3:00 a.m. that very night and inform him the hit failed. tony, manny and chi chi visit frank at his car dealership back office, who is with det. bernstein. nick calls frank, who confirms his involvement by playing the call off as elvira telling him she will be late home. when it becomes apparent that bernstein (who is armed) will not help him, frank begs for tony is forgiveness, saying that he can have elvira and ten million dollars in exchange for sparing his life. tony will have none of it, and manny coldly executes frank. bernstein insists that he could be a valuable ally for tony, but tony disagrees, and kills him too.his problems apparently solved, tony begins a profitable relationship with sosa, marries elvira, buys a new mansion complete with surveillance cameras and numerous luxury items, and tony even sets gina up in business with her own beauty salon. manny and gina soon begin a romantic relationship, but they keep it secret from tony who had firmly stated to gina that he does not want her dating anybody.as tony is business grows, so does his cocaine addiction and paranoia, and he begins to spiral out of control... the beginning of the end. his wife, who becomes further addicted to cocaine, becomes bored and emotionally distant. tony is banker informs him that laundering the increasing flow of drug money has become increasingly difficult, so he will be charging higher fees, up to 10%. a jewish mob boss, mel seidelbaum (ted beniades), contacts manny, offering his assistance. however, as they are cleaning out the money, seidelbaum reveals himself to be an undercover cop and arrests tony. after posting a $5 million in bail, tony is corrupt lawyer, sheffield, tells him that although he may get him cleared of the corruption and money laundering charges, tony will probably have to serve at least three years in prison for tax evasion. manny suggests that he take it, as the american prison system is nowhere near as harsh as its cuban counterpart, and the right legal loopholes could trim the sentence down to six months. however, the strung-out tony yells that he would rather die than spend a single day in jail.after hearing about tony is arrest, sosa, not wanting to lose his main distributor, steps in to intervene by offering tony a way out of going to prison. he calls tony back to bolivia where he introduces him to his cocaine &quot;board of directors&quot; a group that includes a sugar baron, bolivia is military chief, and a mysterious american named charles goodson (gregg henry). we assume he is a corrupt cia officer because sosa guarantees that the irs will not be able to send tony to jail. but this help comes at a price. a bolivian journalist is attempting to expose the ongoing corruption in the bolivian government involvement in drug trafficking, and his crusade is beginning to hurt sosa and his partners. sosa will be sending alberto to new york assassinate the journalist, but he needs tony and his crew to provide some extra muscle. tony is clearly disturbed by the assassination since it is against his custom to kill a man whom he sees as a civilian, plus tony has never killed anybody who did not wrong him personally. but seeing no other options, tony reluctantly agrees to help sosa with the hit.in the meantime, tony is marriage with elvira finally ends when after a bitter altercation at a local restaurant, she finally expresses her contempt for him and the lives he had led her on, and walks out of the restaurant, and out of his life. tony, punch-drunk on cocaine, tells the restaurant is other patrons that his existence is necessary since society needs a man like him to call a criminal. tony also informs manny to look after things while he travels to new york on business (but he does not tell him about the sosa assassination deal.)tony, with his henchmen, chi chi and reuben, and alberto travel to new york city and alberto places a bomb under the journalist is car in with the intention of detonating it outside the un building before the journalist addresses the general assembly and exposes sosa is cartel. but tony has second thoughts when the journalist unexpectedly picks up his wife and children. tony, saying that the team was only supposed to kill only the journalist, shoots alberto to prevent the journalist is family from being killed. when authorities later discover the unexploded bomb underneath the journalist is car, they realize that an execution had been planned and increase the amount of security protecting the journalist. sosa is now the primary suspect and sosa vows to get even with tony.returning to miami, tony discovers that gina and manny (who opposed the trip to new york) have disappeared. tony has long harbored an apparent unnatural obsession for his sister and is overly protective of her for reasons that he may not understand himself. tony visits his mother again where she angrily tells him about gina is descent into an immoral life and accuses him of corrupting her with his flashy lifestyle. after getting gina is home address from mrs. montana, who does not know who else lives there, tony goes to the house in nearby palm grove. much to tony is surprise, manny unexpectedly opens the door in his bathrobe. tony then sees gina in a night gown at the top of the stairs. enraged that another man has obviously slept with his sister, tony shoots manny dead. hysterical, gina reveals to tony that they had just been married and were going to surprise him. tony, riddled with guilt, has gina taken back to his mansion.in revenge for tony is failure to kill the journalist, who has now exposed sosa and his partners to the world as drug lords, sosa sends a latino mercenary hit squad (the size of a large platoon), to tony is mansion to kill him that very evening. sitting at his desk snorting from an enormous pile of cocaine, tony realizes and regrets what he has done to his best friend. when tony is contemplating his actions, sosa is mercenaries breach the main gate at tony is estate and quietly begin to kill all the guards around the mansion. at the same time, a distraught gina, wearing only an unbuttoned sleep shirt and armed with a revolver, enters tony is office to confront him with the truth about his feelings for her. she now realizes that tony loves her in an unnatural way and demands, at gunpoint, that he make love to her. she begins to shoot at him while demanding he take her. a sosa assassin hiding on the balcony, thinking gina is shooting at him, leaps in and riddles her with bullets. an enraged tony throws the man off the balcony and kills him with his sub-machine gun creating a storm of chaos at the mansion. at this point, the mercenaries, robbed of the element of surprise by the gunshots, swarm in to attack tony is mansion from all directions.as all his men are being killed, tony, still delirious from the cocaine, leans over gina is dead body begging for forgiveness, at the same time the mercs break into the mansion, chi chi opens fire with an uzi as he falls back and ends up banging on the door to tony is office (it has been locked from the inside by gina who was planning to kill tony). unfortunately, tony does not seem to hear him. chi chi is shot in the back and tony sees it on the security cameras.as the hit men prepare to storm his office, tony finally snaps out of his drug-induced state, arms himself with an ar-15 assault rifle with an under-mounted m203 grenade launcher and blows down the door. a huge climatic gun battle begins as tony takes position atop the grand staircase and guns down dozens of sosa is men who try to storm the balcony. tony is hit a number of times by return fire, but he keeps shooting. with most of sosa is men dead, tony, strung-out on drugs, defiantly yells out at the assassins, not realizing that the skull has sneaked into the room behind him. the skull shoots tony in the back with a 12-gauge shotgun. tony falls off the balcony and into a reflecting pool at the base of the grand staircase. in the final shot, as the skull and the few surviving assassins look on, tony montana lies dead... face down in the reflecting pool which is located below a large brass globe that says: the world is yours.&#39;, &#39;george falconer (colin firth) approaches a car accident in the middle of a snow-white scenery. there is a bloodied man there and he kisses him. he wakes up: he was dreaming about the moment when his partner of 16 years, jim (mathew goode), died--though he was not there with him because jim was visiting his disapproving family on his own. george remembers the phone ringing on that fateful day, when jim is cousin told him about the fatal accident, and how george was not welcome to attend the funeral, because of the family is homophobia (common for the period and later). george remembers breaking down to charley (julianne moore) that day, his best friend from his life in london, who had also relocated to la; once briefly sexually attached to george before he was completely honest with himself, she may still feel attracted to him.george showers and dresses. it is november 30, 1962, the eve of the cuban missile crisis. though british, he is now a professor of english at ucla. he is depressed, never having recovered from his loss; and when he leaves for work, he packs a gun in his briefcase.he tells his cleaning lady alva (paulette lamori) that she has always been wonderful - in spite of her having forgotten to take out the bread from the fridge. george hugs her, which leaves her utterly confused.on campus, george notices a couple of students, chain-smoking lois (nicole steinwedell) and a boy. one of the secretaries (keri lynn pratt) tells him that she has given his address to some nice new student; it turns out to be this boy, kenny potter (nicholas hoult), who talks to him after class about the speech george has just given out in the classroom concerning minorities and fear. kenny discusses recreational drug use with kenny who tells him that he had never heard george express himself so openly in class as he had that day. he buys george a pencil sharpener as a token of gratitude for george is talking with him.george phones charley, who is dressing for the dinner they have planned at her home. george gets into his car, and picks his gun after having cleaned up his office. however, kenny appears once again, and invites him to go for a drink, observing george is depression and having noticed that he has cleaned out the desk in his office. george tells him it will have to be some other time. he goes to the bank to pick up various things from his safe deposit box, and when looking at a photo of his deceased lover, recalls a conversation with him on the beach.after buying some bullets, he goes to a convenience store. there, carlos (jon kortajarena) bumps onto him, breaking the bottle of scotch he has just bought. george buys a new bottle of scotch and they talk. they smoke a few cigarettes and drink a bottle of gin together. george leaves, refusing carlos\&#39; offer of company, saying that this is a serious day for him and that he is trying to get over an old love.at home, he puts on a record and remembers a conversation with jim while each one was reading a different book on a couch. he pretends shooting himself as practice for later that night, but in a semi-comic scene, can not find the best position in which to accomplish it. charley calls to remind him of their dinner plans, which he grudgingly attends after leaving a note and some money for alva. they dance and talk about london, life, charley is ex-husband is abandonment, and she offends george by suggesting that they might have had a &quot;normal&quot; life together if he had not been a &quot;poof.&quot; charley says george does not look well, reminding him of the heart attack he suffered near the time of jim is death. charley tries to convince george to spend the night at her home, but he leaves.the scene flashes back to 1946 when jim and george had met when at a bar. jim was on leave from the army, right after the second world war. returning to1962, we see george returning to the same bar, near his home; now a quiet place where he asks for a scotch.kenny has followed him there. they talk and then go to the beach and swim naked. they go to george is place. as george is forehead is bleeding, kenny tends to it, and sees in the medicine is cabinet a nude photo of jim. george sees kenny strip off his wet clothes, but does nothing. kenny says that he and lois are not romantically involved. not unlike george and charley in the distant past, kenny explains that they had a brief sexual liason. kenny and george do not have sex, and kenny stays on the couch, given the very late hour.george wakes in a few hours, and finds his gun under kenny is covers and removes it, locking it up as kenny sleeps. when he returns to bed, george dies of a heart attack, seeing the image of jim kissing his forehead.&#39;, &#39;baise-moi tells the story of nadine and manu who go on a violent spree against a society in which they feel marginalized. nadine is a part-time sex worker, and manu a slacker who does anything—including occasional porn film acting—to get by in her small town in southern france. one day, manu and her friend, a drug addict, are accosted in the park by three men, who take them to a garage and gang-rape them. while her friend struggles, screams, and fights against the rapists, manu lies still with a detached look, which troubles the man raping her, who soon gives up. when her friend asks manu how she could act so detached, she replies that she &quot;can not prevent anyone from penetrating her pussy&quot;, so she did not leave anything precious in there. manu then returns to her brother is house, and does not tell him what has happened, but noticing bruises on her neck, he realizes. he gets out a gun and asks manu who was responsible, but when she refuses to tell him, he calls her a &quot;slut&quot; and implies that she actually enjoyed being raped. in response, picking up the discarded pistol, manu shoots him in the head. meanwhile, nadine returns home and has an argument with her roommate, whom she strangles and kills, before leaving with their rent money. nadine suffers another emotional setback when she meets her best friend, a drug dealer, in another town, but he is shot and killed while out obtaining drugs with a prescription she forged for him. later that night, having missed the last train, nadine meets manu at the railway station. manu says she has a car, if nadine will drive for her. they soon realize that they share common feelings of anger, and embark on a violent and sexually charged road trip together. in need of money, the girls hold up a shop and also kill a woman at a cash machine. in a stolen car, they are pulled over for a random check by police, whom they kill. another woman, who was also being checked and saw the murders, flees with them. the women stay over at their new friend is house, whose brother provides the address and details of an architect with whom he has had trouble. the women trick their way into the architect is house and kill him. finally, after this spree of murder and sexual activity, the two women enter a swingers\&#39; club. one of the patrons makes a racist comment to manu. the women kill most of the patrons there, and use a gun to anally penetrate the racist man, finally shooting him. the pair discuss what they have done, and agree that it has all been pointless because nothing has changed in them. during their spree, the duo is crimes are reported by the press, and become a point of fascination for the entire country, with some people actually supporting them, and others afraid of them. when manu enters a roadside tire shop to get some coffee, she is shot by the shop owner, who is then shot by nadine outside. nadine takes manu is body to a forest and burns it, before driving to a beach. with tears in her eyes, nadine puts the gun to her head, intending to commit suicide, but gets arrested by the police before she can do so.&#39;, &#39;kyle pratt (jodie foster) is a propulsion engineer based in berlin, germany. her husband david (john benjamin hickey) died from falling off the roof of an avionic manufacturing building, and now kyle and her six year-old daughter julia (marlene lawston) are flying home to long island to bury him and stay with kyle is parents. they fly aboard a passenger aircraft, an elgin 474, which kyle helped design. after falling asleep for a few hours, kyle wakes to find that julia is missing. after trying to remain calm at first, she begins to panic, and captain marcus rich (sean bean) is forced to conduct a search. kyle walks the aisles, questioning people, but none of her fellow passengers remember having seen her daughter. one of the flight attendants calls in to the airport they just departed from and, shockingly, the gate attendant says that they have no record of julia boarding the flight. in addition, according to the passenger manifest, julia is seat is registered empty. when kyle checks for julia is boarding pass, it is missing.marcus refuses to allow the cargo hold to be searched because the searchers could be hurt if the plane shifted due to turbulence. both marcus and the other crew members suspect that kyle has become unhinged by her husband is recent death, and has imagined bringing her daughter aboard. one flight attendant stephanie (kate beahan) is exceptionally unsympathetic. faced with the crew is increasing skepticism regarding her daughter is existence, kyle becomes more and more desperate. because of her increasingly erratic, panicked behavior, air marshal gene carson (peter sarsgaard) is ordered by marcus to guard and handcuff her.later on, marcus receives a wire from the hospital in berlin. it says that julia was with her father when he fell off the roof, and she also died of internal injuries. kyle furiously denies that, consistently claiming that she brought julia aboard. the crew now believes she is delusional and she is brought back to her seat by air marshal carson. a therapist (greta scacchi) on board tries to console her, causing kyle for a moment to doubt her own sanity until she notices that a heart julia had drawn earlier on the window next to her seat is real. kyle is emboldened and convinces the therapist to let her use the bathroom. with her handcuffs now removed, she climbs into the upper compartment and sabotages the aircraft is electronics, deploying the oxygen masks and interrupting lighting. some passengers brawl with the red herring arab passengers on board. she uses the chaos to take an elevator to the lower freight deck. she desperately searches for julia and finally opens her husband is casket to which she emotionally breaks down. carson finds her and escorts her back.kyle makes a final plea to carson that she needs to search the plane upon landing. carson considers for a moment, then &quot;goes to speak to the captain,&quot; against flight attendant stephanie is command (they are landing), leaving the audience to momentarily believe he is sympathetic. instead, he sneaks back into the freight deck to remove two small explosives and a detonator concealed in david is casket. he then climbs down to a part of the avionics section, revealing julia is sleeping (presumably drugged) with her coat and backpack that no one could find. he attaches the explosives to the side of the platform and arms them. at this point, it is revealed that carson, stephanie, and the coroner in berlin (christian berkel) are the antagonists and part of a conspiracy. carson tells the captain that kyle is a hijacker and is threatening to blow up the aircraft with explosives hidden in the un-x-rayed casket unless the airline transfers $50,000,000 into a bank account. it is revealed that the conspirators killed kyle is husband and abducted julia in order to frame kyle. carson tells an unnerved stephanie that he intends to blow up the aircraft is avionics section, killing the unconscious julia, and leave kyle dead with the detonator in her hand.after making an emergency landing at goose bay airport in goose bay, labrador, the passengers exit the aircraft as the tarmac is surrounded by u.s. federal bureau of investigation (fbi) agents (though the airport is properly royal canadian mounted police jurisdiction). as the captain is leaving, kyle runs to speak to him with carson in tow. the captain demands she give up her charade, revealing carson is deception. quickly playing the role of hijacker, kyle demands that carson stay on board and the crew disembark.as soon as the plane is door closes, kyle knocks carson unconscious with a fire extinguisher, handcuffs him to a rail, and takes the detonator from his pocket. stephanie comes out of hiding and kyle screams &quot;she is in avionics is not she?&quot; carson quickly regains consciousness and fires at kyle with a concealed gun, sending her running. he chases after kyle shooting, until she locks herself in the cockpit. he reveals his conspiracy to talk her out. she opens a hatch door to the upper level and throws out a binder to fool him. carson hears the upstairs thud and leaves. kyle exits and encounters a guilt-ridden stephanie slapping her palm with a large flashlight. kyle talks her down and punches her out. stephanie panics and flees the plane, abandoning carson who looks on.kyle during this time searches avionics and finally finds the unconscious julia. carson soon follows, and while searching, tells her how he gagged and dumped her daughter into the food bin. he disparages the people aboard who would never care enough to notice. carson points his gun to where julia lay before, but they are not there. he turns around and sees kyle carrying julia, escaping through a small door with the detonator in hand. carson shoots at her as she closes the door. she detonates the explosives, killing carson. the compartment she and julia hid in was non-combustible, which keeps them safe. kyle, carrying julia, exits via a cargo door. everyone watches in shock and amazement as kyle carries her daughter out onto the tarmac.later in the passenger waiting section of the airport, marcus apologizes to a seated kyle holding julia in her arms. stephanie is led away by fbi agents and more agents approach kyle, asking her to identify the morgue director in berlin who has been detained. she carries julia still unconscious through the crowd of passengers, silently redeeming herself. one of the arab passengers (michael irby) helps pick up her bag, a symbolic act of respect and forgiveness. before loading her daughter into a van to take them away, julia wakes up and sleepily asks &quot;are we there yet?&quot; as they drive away. in this movie appears christian berkel who will later play the role of fritz shimon haber in haber.&#39;, &#39;small-time italian-american criminals caesar enrico &quot;rico&quot; bandello (edward g. robinson) and his friend joe massara (douglas fairbanks, jr.) move from new york to chicago to seek their fortunes. rico joins the gang of sam vettori (stanley fields), while joe wants to be a dancer. olga (glenda farrell) becomes his dance partner and girlfriend at the local taxi dance club.joe tries to drift away from the gang and its activities including running several speakeasys and illegal gambling casinos, but rico (whom the gang now refers to by his nickname \&#39;little caesar\&#39;) makes him participate in the robbery of the nightclub where he works. despite orders from underworld overlord &quot;big boy&quot; (sidney blackmer) to all his men to avoid bloodshed, rico guns down crusading police crime commissioner alvin mcclure during the robbery, with joe as an aghast witness.rico accuses sam of becoming soft and seizes control of his organization. rival boss &quot;little arnie&quot; storch (maurice black) tries to have rico killed, but rico is only grazed by a bullet during a drive-by shooting. rico and his gunmen pay little arnie a visit, after which arnie hastily departs for detroit. the big boy eventually gives rico control of all of chicago is northside.some months later, rico becomes concerned that joe knows too much about him. he warns joe that he must forget about olga, and join him in a life of crime. rico threatens to kill both joe and olga unless he accedes, but joe refuses to give in. olga calls police sergeant flaherty and tells him joe is ready to talk, just before rico and his henchman otero (george e. stone) come calling. rico finds, to his surprise, that he is unable to take his friend is life. when otero tries to do the job himself, rico wrestles the gun away from him, though not before joe is wounded. hearing the shot, flaherty and another cop give chase and kill otero. with the information provided by joe, flaherty proceeds to crush rico is organization.desperate and alone, rico retreats to the gutter from which he sprang. a few weeks later, while hiding in a flophouse, he becomes enraged when he learns that flaherty has called him a coward in the newspaper. he foolishly telephones the cop to announce he is coming for him. the call is traced to the phone booth where rico is. he runs from the police and hides behind a large billboard. refusing to surrender, flaherty personally shoots at the billboard with a tommy gun. ironically, the billboard shows an advertisement featuring dancers joe and olga. the police walk around the billboard to find rico dying on the ground who with his last breath mutters, &quot;mother of mercy... is this the end of rico?&quot;&#39;, &#39;the movie begins with a video being shot of men with their hands tied behind their backs. an enforcer, who we learn later is lado (benicio del toro) wears a skull lucha libre mask and runs a chainsaw.next, there is a shot of an attractive young blonde woman named \&#39;o\&#39; (short for ophelia, played by blake lively) as she narrates in a voice-over. the love of her life is two men: ben (aaron johnson) and chon (taylor kitsch). to her, ben is like warm wood, the idealistic buddhist; chon is steel, the baddhist. o is in a polyamorous relationship with both men, and the three live together in laguna beach, california in a luxurious cliff-side house for which ben paid with cash. ben and chon are high school buddies. ben went to berkeley and majored in business and botany. chon was a navy seal with tours of duty in iraq and afghanistan. one day, lounging on the beach, ben asks chon where you can get the best weed. chon replies afghanistan, and during a tour there, smuggles out hashish/marijuana seeds from which ben grows in a greenhouse, using his scientific knowledge, cannabis with a thc yield of 33%, &quot;the best in the world.&quot; they met o when they hired her to help with managing the greenhouse. ben is the businessman who runs the production and distribution and handles 99% of the business, while chon handles the 1% (enforcement). ben is philosophy is to provide what people want, but chon is the realist in seeing the money made from marijuana. ben does a lot of humanitarian work, and sells to marijuana clubs and dispensaries, for medical usage, but most of the money is made from selling on the street, where it goes for about $6,000 a pound.after making love to o, chon sees on his computer an e-mail. the e-mail has a video of lado kicking around the decapitated heads and reads: &quot;these guys were stupid. be smarter we need to talk.&quot; o and chon greet ben returning from a volunteerism trip to africa. chon and ben discuss the business. ben wants to give up the business to the mexican cartel as he wants to go into philanthropic ventures such as making $10 solar panels, and he admits he is afraid of the cartel. but chon argues that selling would show weakness, and once they do this, the cartel would usurp their whole lives.we see lado and his \&#39;gardening crew\&#39; carry out a hit. lado accosts a middleman (shea whigham) in his home. the middleman pleads forgiveness and promises an even better deal than paying up. he calls el azul (joaquin cosio) (spanish for the blue), a cartel head honcho, to connect him with lado, but lado shoots the middleman and has a teenage cartel-member-in-training, esteban, shoot the middleman is wife, while el azul is on the phone, enraging him. the gardening crew smuggles the bodies out of the house in trash bins. we learn that el azul is a former inside man with the baja cartel, the cartel lado works for. the head of the baja cartel, elena sanchez (salma hayek), had a falling-out with el azul, and now el azul is encroaching on the baja cartel is territory, threatening to take over her position.meanwhile, ben and chon meet with dennis (john travolta), a corrupt dea agent on ben and chon is payroll. dennis tells them about the baja cartel: they want ben and chon is market to fill a market they do not have. dennis urges them to take the deal instead of decapitation; he reminds them that times are changing and they need to adapt. dennis is sympathetic to them because his wife has cancer and ben provides him with cannabis to ease the effects of her chemotherapy.ben and chon go to a hotel room and meet with alex (demian bichir), the representative for elena. the meeting is secretly video-recorded by lado, with the video feed shown to elena in her tijuana compound.alex lays out the deal: a contract of three years of working together, during which the cartel will learn ben is horticulture methods while providing vast distribution networks and the operation in a safe haven on an indian reservation that the cartel owns. (the indian reservations are considered out of u.s. jurisdiction.) during the meeting, chon badmouths the cartel is products by saying &quot;i think, basically, you want us to eat your shit and call it caviar.&quot; ben and chon step out into balcony to discuss. we see that a formidable security apparatus, a team of chon is ex-navy seal friends, who have sniper wares trained on the hotel room, backing them. chon had planned to take out alex and his men, but ben talks him out of it. ben and chon reach a decision. they will turn over their business to the cartel, but not join them. the cartel is not happy. (apparently, no one says \&#39;no\&#39; to the baja cartel.)fearing the consequences of their decision, ben meets with his money guy, spin (emile hirsch), an ex-goldman sachs banker who launders and invests the marijuana company is money. ben wants to set up funds to disappear for three years in indonesia. ben, chon, and o have dinner and discuss the plan for their new lives. unbeknownst to them, they are being watched in the restaurant by lado is crew. the three celebrate by getting high in their beach house and having a threesome. lado sits in the car, spying on them. he calls elena to update her on the situation. elena says she needs ben and chon to join her for their expertise. hearing about their living situation, she found their weakness: o.the next day, the three prepare to leave for indonesia. o wants to do a last minute shopping trip and chon assigns someone from the security team for her. o is driving to south coast plaza (an upscale mall in orange county), but one of lado is men intercepts the bodyguard and kills him. they tail o. in one scene, o is shopping next to a girl who is talking on a cellphone, answering her mom is questions about college life. her mom turns out to be elena.returning home but not finding o, ben and chon hear a chime indicating a skype call, and to their horror, see a live web feed of the kidnapped o. elena, her voice masked, reprimands the boys for not complying with the deal. to redress the insult of the cartel is product during the meeting, elena has chon put a gun into his mouth, indicating she will tell lado to cut off two of o is fingers otherwise. she almost gets chon to pull the trigger but stops him at the last minute, telling him this will clean up his mouth.elena tells ben and chon to deliver 300 pounds of their marijuana in five hours to show compliance. they get the product and race to chula vista. they arrive just in time despite being almost pulled over (and chon almost having to kill the highway patrolman). lado and his men take their car. ben and chon drive back in lado is truck, which contains the suitcase of money (they paid in full, ben and chon note) and a cell phone. elena calls, her voice unmasked, and tells them that she will hold o hostage for one year to make sure they comply with all of her demands. chon argues with ben that they have to fight to get o back -- the cartel members are savages who will kill them anyway. but ben just wants peace and o back.ben and chon meet again with dennis, who tells them about the conflict between elena and el azul. chon strong-arms dennis into getting data on elena and the baja cartel from the dea. he is able to get 60 days of financial transactions, encoded onto two credit cards. dennis also tells them that two of elena is sons and her daughter (magdalena), are dead. back in their company hq, ben and chon have spin and his experts look into the financial data on the computer. from it, they trace elena is financial routes. they figure out that one of her real-estate holdings, a trailer house, is a drop off point for the cartel is business. ben decides on a plan to pay to get o back. he and elena negotiate to a $13 million ransom. they decide to hijack the drop to get the cash for o is ransom, and make it look like el azul hit the cartel.the hijacking gets underway. wearing masks, chon and his team of ex-navy seals blow up the front car in a convoy with ieds. when the men from the trailer retaliate, an ex-navy seal member takes them out with rpg rockets. chon forces the men out of the car carrying the cash by putting a grenade in the gas tank. he takes the briefcase from a man scrambling out of the car. ben is horrified by the killings. one of the cartel members is about to shoot ben, but chon shoots the member in the back, spraying blood on ben is mask. they take the money and get away. alex and lado arrive to assess the damage, which is 3 million dollars gone and seven men dead. lado urges alex not to inform elena of the loss because he wants to recover the money and rectify the situation himself and because he fears elena.ben and chon get a message to have ben meet with alex and lado. alex and lado interrogate ben to find out their whereabouts during the hijacking but ben plays it cool. he insinuates that it could be lado is men. after ben leaves, alex decides ben and chon do not have the balls for the attack. he suspects el azul but lado is not so sure.we see o at the shack where they are holding her. she is being watched over by esteban. her living conditions are bad: one of her legs is chained to a chair, she has a bucket for bathroom and sleeps on a spring bed in a cage. she pleads into the surveillance camera and gets to talk to elena, and begs her for some concessions. elena asks o how long she has been using. since 8th grade, o says, not that her mother cared. later we see lado feeding o a steak, and later, giving o a drug fix, and her getting high. o is also given access to a laptop to send a fabricated email to her mother so she will not file a missing persons report on o. afterwards, lado tells esteban, do not look at me like that (because we see later in the movie lado shows o a cellphone video of him raping her while she was high).elena is daughter, magdalena (sandra echevarria) is seen having sex when she is interrupted by a phone call. her mother is on the line, informing her that she is coming to california to visit. elena gets on a private plane to the indian reservation in order to avoid u.s. customs. lado will follow her later, bringing o with him. before boarding the plane, lado tells esteban that he is proven too sensitive to handle the work and proceeds to shoot and kill him.upon arrival incalifornia, lado brings o to elena. elena immediately informs lado that she knows about the hijacking, and chastises him for not telling her. she threatens lado, telling him to recover the money and he informs her that he did not tell her because he was trying to get proof that ben and chon were behind the hijacking.ben has spin fabricate documents and hack into alex is bank accounts to frame alex as the mole to el azul. ben confronts lado with alex is dossier in an effort to take the heat off of him and chon. to save himself, lado visits el azul and makes a deal to work with him.at a posh spa, elena and o have dinner during which we learn that elena inherited the business after her husband was killed. her twin sons and brothers are also killed, leaving two children who resent her and do not want her in their lives (her daughter and a third son). elena tells o she likes her, but will not hesitate to kill ben and chon. when o remarks that only she can \&#39;get ben and chon together,\&#39; elena scoffs, telling o that her love story has a serious problem: ben and chon must love each other more than o because otherwise, they would not share her.later, lado brings ben and chon to see alex being whipped. the torture is video-fed to elena who forces o to watch once he \&#39;confesses.\&#39; under the threat of his family being harmed, alex confesses to working with el azul. he begs elena for forgiveness, citing the names of her husband and children. lado douses alex with gasoline and anoints ben to be the executioner. to go through with the falsehood, ben ignites alex on fire. o watches in horror.the next morning, ben and chon walk along the beach where they discuss the night is events. chon tries to justify what ben did by referring to the dalai lama, but ben brushes it off. instead, he tells chon that during alex is plea to elena, alex swore on magda is life; meaning that elena is daughter is alive, and that dennis knew this and hid it from him and chon. they make a plan that will work even faster than accumulating the money for ransom: kidnap elena is daughter and trade her for o.lado pays dennis a visit at his home. it turns out lado also has dennis on the payroll. lado questions dennis on the culprits behind the hijacking and dennis assures him it was azul. lado informs dennis that it can not be azul because he is working with him now instead of elena. under threat, dennis reveals that the hijackers are ben and chon, but that they are going to give the $3 million to elena and not lado. to head off elena, dennis tells lado to kidnap magda and sell her to ben and chon. dennis is incredulous to learn that elena is on u.s. soil.dennis meets with ben and chon, who chew him out for giving them away to lado. the boys offer dennis the $3 million they stole in exchange for magdalena is address. they kill her bodyguard, who is sitting outside her dorm in his car, then move inside, tying up her boyfriend and kidnapping her. wearing masks as a show, they skype with elena and show a shivering magda, tied up and stuffed in an ice box. elena meekly agrees to their demands. at the end of the talk, o informs elena that she knew her boys would come for her. elena slaps o, then screams at everyone to get out. she drops to the ground, rips off her wig, and breaks down.lado and his boys take o into another room, where she is chained to a chair. lado makes everyone else leave the room, then he shows her the video he took when he raped her while she was drugged. she spits in his face, but he enjoys it, wiping himself off on her hair.elena is suv and ben and chon is suv meet in an open field with their hostages. on the mountain ridges, each group has their sniper team in position. elena steps into the open with o and releases her. o runs to ben. ben and chon release magda to elena, who is told by her daughter that she will never speak to her again. before elena returns to her car, she asks chon who the rat is. chon smiles and nods toward lado, who blows back a kiss. elena turns around to shoot lado, but lado draws his gun and shoots her first. o, who was handed a gun by ben, makes her way to elena and wants to kill her, but relents when she sees elena is already dying. elena begs o to help magda escape. o glares at magda briefly before hustling her to safety.a massive shootout is underway. each group is sniper team shoots at each other, taking out some members of each team. chon kills elena is bodyguards and ben and chon converge on lado, hiding behind the suv. ben shoots lado three times in the back, but lado turns around to shoot ben in the throat. lado sprawls on the ground, severely wounded. o shoots lado dead for revenge, and erases his phone. o rushes to ben, seeing his wound is lethal. as o cradles ben, chon pulls out a syringe and injects ben, o, and then himself with drug overdoses. they die in a triple suicide. &quot;that is the perfect ending,&quot; o narrates while looking into the sky as she dies.but wait... rewind, as o mentions that this was just how she imagined it.what really happened, as o narrates, &quot;was more of a fuck-up than a shootout.&quot; after elena steps out into the open and releases o, lado retreats to the car and drives away. a helicopter and dea agents led by dennis swoop in to arrest elena. the dea agents disarm the sniper teams and arrest everyone (including ben, chon, and o).o narrates as we see a montage: dennis is grandstanding on tv about elena is arrest and the progress on the war on drugs (indian grounds are still u.s. grounds he says); elena goes to prison for 30 years, while ben and chon, who have enough information to put dennis away, are released after a few weeks in a holding cell after dennis said that they were confidential informants. lado and el azul form a new cartel to take over the vacuum left by elena.after their release, o says that she, ben, and chon all vanished. montages are shown of them living new lives in indonesia, africa, and other parts of the world. a few scenes show that they continue to remain in touch and get together, but most of the scenes strongly imply that for the most part, they have all gone their separate ways-- as o narrates, perhaps it is true that three people can not all be equally in love. she says that one day, perhaps they will return to their pot business in laguna, but for now, they live simple, even primitive lives, like \&#39;beautiful savages.\&#39;&#39;]] is not valid. Should be a string, a list/tuple of strings or a list/tuple of integers.

In [ ]:
import time
start_time = time.time()

In [ ]:
#For Type 1 Embeddings
xlnet_embeddings_t1=getXLNetEmbeddings(mpstDF_processsed,"processed_synopsis_t1")

In [ ]:
print("Shape: ",xlnet_embeddings_t1.shape)
print("XL Embedding for Type 1")
print(xlnet_embeddings_t1)

In [ ]:
np.savez("xl_embeddings_type1.npz",xlnet_embeddings_t1)

In [ ]:
#For Type 2 Embeddings
xlnet_embeddings_t2=getXLNetEmbeddings(mpstDF_processsed,"processed_synopsis_t2")


In [ ]:
print("Shape: ",xlnet_embeddings_t2.shape)
print("Embedding")
print(xlnet_embeddings_t2)

In [ ]:
np.savez("xl_embeddings_type1.npz",xlnet_embeddings_t2)

Saved the embeddings in different files
Now saving them in the same file

In [ ]:
np.savez("xl_embeddings.npz",t1=xlnet_embeddings_t1,t2=xlnet_embeddings_t2)

In [ ]:
em_check=np.load("xl_embeddings.npz")
print("t1")
print(em_check["t1"])
print(em_check["t1"].shape)
print("t2")
print(em_check["t2"])
print(em_check["t2"].shape)


In [ ]:
print("--- %s seconds ---" % (time.time() - start_time))